files/system/usr/libexec/setup-luks-fido2-unlock

```diff
@@ -0,0 +1,89 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+# The script below has been modified by secureblue to change function to do fido2 unlock
+
+set -eou pipefail
+
+[ "$UID" -eq 0 ] || { echo "This script must be run as root."; exit 1;}
+
+echo "WARNING LUKS drive encryption must have been enabled at install time for this script to run" 
+echo "ENSURE you save the backup key this script creates at /var/home/$SUDO_USER/Desktop/luks_backup_key.txt ON ANOTHER COMPUTER"
+echo "This script uses systemd-cryptenroll to enable FIDO2 auto-unlock."
+echo "You can review systemd-cryptenroll's manpage for more information."
+echo "If you are using usbguard, plug into your hardware key, run 'usbguard list-devices'. Identify which number on the left is"
+echo "your device then run 'usbguard allow-device <number> -p'. You must exit this script with ctrl-C and do this now, BEFORE proceeding"
+echo "WARNING this script is designed not to, but could clear stored secrets on your fido2 key. Ensure you have backup options."
+read -p "Are you sure are good with this and want to enable FIDO2 auto-unlock? (y/N): " -n 1 -r
+echo
+if [[ ! $REPLY =~ ^[Yy]$ ]]; then
+  [[ "$0" = "${BASH_SOURCE[0]}" ]] && exit 1 || return 1 # handle exits from shell or function but don't exit interactive shell
+fi
+
+## Inspect Kernel Cmdline for rd.luks.uuid
+RD_LUKS_UUID="$(xargs -n1 -a /proc/cmdline | grep rd.luks.uuid | cut -d = -f 2)"
+
+# Check to make sure cmdline rd.luks.uuid exists
+if [[ -z ${RD_LUKS_UUID:-} ]]; then
+  printf "LUKS device not defined on Kernel Commandline.\n"
+  printf "This is not supported by this script.\n"
+  printf "Exiting...\n"
+  exit 1
+fi
+
+# Check to make sure that the specified cmdline uuid exists.
+if ! grep -q "${RD_LUKS_UUID}" <<< "$(lsblk)" ; then
+  printf "LUKS device not listed in block devices.\n"
+  printf "Exiting...\n"
+  exit 1
+fi
+
+# Cut off the luks-
+LUKS_PREFIX="luks-"
+if grep -q ^${LUKS_PREFIX} <<< "${RD_LUKS_UUID}"; then
+  DISK_UUID=${RD_LUKS_UUID#"$LUKS_PREFIX"}
+else
+  echo "LUKS UUID format mismatch."
+  echo "Exiting..."
+  exit 1
+fi
+
+# Specify Crypt Disk by-uuid
+CRYPT_DISK="/dev/disk/by-uuid/$DISK_UUID"
+
+# Check to make sure crypt disk exists
+if [[ ! -L "$CRYPT_DISK" ]]; then
+  printf "LUKS device not listed in block devices.\n"
+  printf "Exiting...\n"
+  exit 1
+fi
+
+## Run crypt enroll
+echo "Enrolling FIDO2 unlock requires your existing LUKS unlock password"
+systemd-cryptenroll --fido2-device=auto "$CRYPT_DISK"
+sed -i '$ s/$/ fido2-device=auto/' /etc/crypttab
+
+echo "Creating backup key"
+systemd-cryptenroll --recovery-key "$CRYPT_DISK" > "/var/home/$SUDO_USER/Desktop/luks_backup_key.txt"
```

Comment:
it seems like we have duplicate recovery options. In that, this script doesn't remove the passphrase unlock, and also creates an independent recovery key. I'm not saying we should remove the passphrase unlock, but maybe the recovery key should be optional? Does the user need the recovery key if they're not going to be removing the passphrase backup?

Reply:
If they then run the TPM unlock remove script (assuming they ran the TPM unlock luks script as we recommend), it will remove the passphrase slot as it's attached to the TPM credentials. Considering that the limit is 32 unlock slots, I don't see any harm in being **absolutely** sure that there is another option.

Reply:
good point